### DISCUSSION TITLE
Ensure component storage is fully contiguous?

### DISCUSSION THREAD

**Discussion by benpm** [skypjack/entt#982] [Q&A] [answered]
Hi all, I've been searching around but I was unable to find a clear answer to this question: how can I ensure that all components of a specified type are stored fully contiguously (tightly packed / one page)? From what I've been able to gather, it seems like setting the `page_size` component trait to something extremely large is the only straightforward way to do this, but that seems like a bad hack.

**Comment by skypjack** [comment, marked as answer]
There was (actually, there is) a plan to make it fully contiguous when the page size is set to max value.
However, this is somewhat a low priority task at the moment. Therefore I've not an eta for this feature yet.
Using a page size that is large enough works and avoid reallocations. The latter are one of the main issues with fully contiguous storage.
The other option is to just customize the storage type for `T` and make it use an `std::vector` internally or any other container. As long as the API is the expected one, the rest of the library and likely of your codebase won't be affected.
EnTT is designed to support mixins and full customizations of the storage types. I'm pretty sure it won't complain in case. 🙂

**Reply by benpm** [reply]
Thanks, that's exactly what I need. How would I go about customizing the storage type for a particular component?

**Reply by skypjack** [reply]
Oh, that's easy actually. Something along this line should work:
```cpp
template<typename Entity, typename Allocator>
class basic_storage<Entity, MyCustomType, Allocator> {
    // ...
}
```
Then you need to implement a number of functions to make it work with views (like `get` and `get_as_tuple`) and override the internal virtual functions of interest (such as `pop` or `try_emplace`) to handle your elements properly.
On the `wip` branch, there is a storage customization to create a self-managed entity pool for the registry. You can take inspiration from there or just reach out if you've doubts. 👍
